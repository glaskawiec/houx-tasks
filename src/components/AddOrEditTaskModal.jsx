import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import AddTaskForm from './AddOrEditForm';
import device from '../device';

const AddTaskModalWrapper = styled.div`
  position:fixed;
  background: white;
  width: 80%;
  height: auto;
  top:50%;
  left:50%;
  transform: translate(-50%,-50%);
  border-radius: 0.5rem;
  box-shadow: 0 3px 7px rgba(0, 0, 0, 0.3);
  padding: 2rem;
   
  @media ${device.tablet} { 
    width: auto;
  }
`;

const ModalTitle = styled.h3`
  text-align: center;
`;

const AddOrEditTaskModal = ({ editTaskId }) => {
  console.log(editTaskId);
  const isEditForm = editTaskId !== null;
  return (
    <AddTaskModalWrapper data-testid="add-or-edit-task-modal">
      <ModalTitle>{isEditForm ? 'Edit task' : 'Add new task'}</ModalTitle>
      <AddTaskForm editTaskId={editTaskId} />
    </AddTaskModalWrapper>
  );
};

AddOrEditTaskModal.propTypes = {
  editTaskId: PropTypes.number,
};

AddOrEditTaskModal.defaultProps = {
  editTaskId: null,
};

export default AddOrEditTaskModal;
