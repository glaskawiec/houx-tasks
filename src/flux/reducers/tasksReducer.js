import {
  TASKS_ADD,
  TASKS_REMOVE,
  TASKS_COMPLETE,
  TASKS_EDIT,
  TASKS_OPEN_ADD_MODAL,
  TASKS_OPEN_EDIT_MODAL,
  TASKS_CLOSE_MODAL, TASKS_CHANGE_ORDER,
} from '../actionTypes/tasksActionTypes';

export const initialState = {
  tasks: [
    {
      id: 0,
      title: 'Example task',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
      colors: {
        text: 'pink',
      },
    },
    {
      id: 1,
      title: 'Example task 2',
      description: '',
      colors: {
        text: 'purple',
      },
    },
    {
      id: 2,
      title: 'Example task 3',
      description: '',
      colors: {
        text: 'yellow',
      },
    },
  ],
  nextId: 3,
  isEditModalOpen: false,
  currentEditTaskId: null,
  isAddModalOpen: false,
};

const tasksReducer = (state = initialState, action) => {
  switch (action && action.type) {
    case TASKS_ADD: {
      const newTodo = {
        ...action.task,
        id: state.nextId,
      };

      return {
        ...state,
        nextId: state.nextId + 1,
        tasks: [...state.tasks, newTodo],
      };
    }
    case TASKS_REMOVE: {
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.id),
      };
    }
    case TASKS_COMPLETE:
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.id),
      };
    case TASKS_OPEN_ADD_MODAL:
      return {
        ...state,
        isAddModalOpen: true,

      };
    case TASKS_CLOSE_MODAL:
      return {
        ...state,
        isEditModalOpen: false,
        isAddModalOpen: false,
      };
    case TASKS_OPEN_EDIT_MODAL:
      return {
        ...state,
        isEditModalOpen: true,
        currentEditTaskId: action.id,
      };
    case TASKS_EDIT: {
      const { currentEditTaskId } = state;
      return {
        ...state,
        tasks: [
          ...state.tasks.filter((task) => task.id !== action.id),
          { ...action.editedTask, id: currentEditTaskId },
        ],
        currentEditTaskId: null,
      };
    }
    case TASKS_CHANGE_ORDER: {
      const {orderDown, id} = action;
      const tasksCopy = [...state.tasks];
      let tasksToOrderIndex = 0;

      for( ; tasksToOrderIndex < tasksCopy.length; ++tasksToOrderIndex){
        if(tasksCopy[tasksToOrderIndex].id === id){
          break;
        }
      }

      if(orderDown){
        const previousId = tasksCopy[tasksToOrderIndex+1].id;
        tasksCopy[tasksToOrderIndex].id = previousId;
        tasksCopy[tasksToOrderIndex+1].id = id;
      } else {
        const previousId = tasksCopy[tasksToOrderIndex-1].id;
        tasksCopy[tasksToOrderIndex].id = previousId;
        tasksCopy[tasksToOrderIndex-1].id = id;
      }

      return {
        ...state,
        tasks: tasksCopy
      };
    }
    default: return state;
  }
};

export default tasksReducer;
