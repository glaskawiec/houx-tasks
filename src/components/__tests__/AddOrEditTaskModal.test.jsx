import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { HouxProvider } from 'houx';
import { getNodeText } from '@testing-library/dom';
import AddOrEditTaskModal from '../AddOrEditTaskModal';
import tasksReducer from '../../flux/reducers/tasksReducer';

describe('AddOrEditTaskModal component unit tests', () => {
  const mockedState = {
    tasks: [
      {
        id: 0,
        title: 'Test task 1',
        description: 'Test description 1',
        colors: {
          text: 'pink',
        },
      },
      {
        id: 1,
        title: 'Test task 2',
        description: '',
        colors: {
          text: 'purple',
        },
      },
    ],
    nextId: 1,
    isEditModalOpen: false,
    currentEditTaskId: null,
    isAddModalOpen: false,
  };

  const mockedReducers = {
    tasks: (state, action) => tasksReducer(mockedState, action),
  };

  test('AddOrEditTaskModal should render properly', () => {
    const { container } = render(
      <HouxProvider reducers={mockedReducers}>
        <AddOrEditTaskModal />
      </HouxProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Should title be mandatory during adding new task', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <AddOrEditTaskModal />
      </HouxProvider>,
    );

    fireEvent.click(getByTestId('submit-button'));

    expect(getByTestId('title-input-error')).toBeInTheDocument();
    expect(getNodeText(getByTestId('title-input-error'))).toBe('Cannot be empty.');
  });
});
