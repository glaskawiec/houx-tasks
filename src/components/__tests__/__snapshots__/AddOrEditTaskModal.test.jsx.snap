// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddOrEditTaskModal component unit tests AddOrEditTaskModal should render properly 1`] = `
<div>
  <div
    class="sc-AxhCb bQtECx"
    data-testid="add-or-edit-task-modal"
  >
    <h3
      class="sc-AxhUy fNfuCs"
    >
      Add new task
    </h3>
    <form
      class=""
    >
      <div
        class="form-group"
      >
        <input
          class="sc-AxirZ HlbFI form-control"
          data-testid="title-input"
          error="false"
          id="title"
          placeholder="Title"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          data-testid="description-input"
          id="description"
          placeholder="Description"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          data-testid="background-color-input"
          id="background-color"
          placeholder="Background Color"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          data-testid="text-color-input"
          id="text-color"
          placeholder="Text Color"
          value=""
        />
      </div>
      <button
        class="sc-AxjAm dwDcds btn btn-primary"
        data-testid="submit-button"
        type="submit"
      >
        Add
      </button>
      <button
        class="btn btn-primary"
        data-testid="close-button"
        type="button"
      >
        Close
      </button>
    </form>
  </div>
</div>
`;
